pallets/token-mux/src/lib.rs

```diff
@@ -0,0 +1,328 @@
+				OrderRatio::Market => RatioFor::<T>::ensure_from_rational(
+					amount,
+					T::Swaps::convert_by_market(
+						order.currency_in.clone(),
+						order.currency_out.clone(),
+						amount,
+					)?,
+				)?,
+				OrderRatio::Custom(ratio) => ratio,
+			};
+
+			ensure!(ratio == One::one(), Error::<T>::NotIdenticalSwap);
+
+			match (
+				Self::try_local(&order.currency_out),
+				Self::try_local(&order.currency_in),
+			) {
+				(Ok(_), Ok(_)) | (Err(_), Err(_)) => {
+					return Err(Error::<T>::InvalidSwapCurrencies.into())
+				}
+				(Ok(local), Err(_)) => {
+					ensure!(
+						order.currency_in == local,
+						Error::<T>::InvalidSwapCurrencies
+					);
+
+					T::Tokens::mint_into(local, &Self::account(), amount)?;
+					T::Swaps::fill_order(Self::account(), order_id.clone(), amount)?;
+				}
+				(Err(_), Ok(local)) => {
+					ensure!(
+						order.currency_out == local,
+						Error::<T>::InvalidSwapCurrencies
+					);
+
+					T::Swaps::fill_order(Self::account(), order_id.clone(), amount)?;
+					T::Tokens::burn_from(
+						local,
+						&Self::account(),
+						amount,
+						Precision::Exact,
+						Fortitude::Polite,
+					)?;
+				}
+			}
+
+			Self::deposit_event(Event::<T>::SwapMatched {
+				id: order_id,
+				amount,
+			});
+
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T>
+	where
+		CurrencyFor<T>: From<LocalAssetId> + TryInto<LocalAssetId>,
+	{
+		fn account() -> T::AccountId {
+			T::PalletId::get().into_account_truncating()
+		}
+
+		fn mint_route(
+			who: &T::AccountId,
+			local: CurrencyFor<T>,
+			variant: CurrencyFor<T>,
+			amount: BalanceFor<T>,
+		) -> DispatchResult {
+			T::Tokens::transfer(
+				variant,
+				&who,
+				&Self::account(),
+				amount,
+				Preservation::Expendable,
+			)?;
+
+			T::Tokens::mint_into(local, &who, amount).map(|_| ())
+		}
+
+		fn burn_route(
+			who: &T::AccountId,
+			local: CurrencyFor<T>,
+			variant: CurrencyFor<T>,
+			amount: BalanceFor<T>,
+		) -> DispatchResult {
+			T::Tokens::burn_from(local, &who, amount, Precision::Exact, Fortitude::Polite)?;
+
+			T::Tokens::transfer(
+				variant,
+				&Self::account(),
+				&who,
+				amount,
+				Preservation::Expendable,
+			)
+			.map(|_| ())
+		}
+
+		fn try_local(currency: &CurrencyFor<T>) -> Result<CurrencyFor<T>, DispatchError> {
+			let meta_variant =
+				T::AssetRegistry::metadata(currency).ok_or(Error::<T>::MetadataNotFound)?;
+
+			let local: CurrencyFor<T> = meta_variant
+				.additional
+				.local_representation
+				.ok_or(Error::<T>::NoLocalRepresentation)?
+				.into();
+
+			let meta_local =
+				T::AssetRegistry::metadata(&local).ok_or(Error::<T>::MetadataNotFound)?;
+
+			// NOTE: We could also think about making conversion between local
+			//       representations and variants but I fear that we then have problems with
+			//       SUM(locked variants) = local. Hence, this restriction.
+			ensure!(
+				meta_local.decimals == meta_variant.decimals,
+				Error::<T>::DecimalMismatch
+			);
```

Comment:
I feel like this restriction is hard. I would expect to have different decimal representation of currencies that has a 1:1 relation. But sure @mustermeiszer can say more here

Reply:
The purpose of this pallet is to aggregate variants of the same currency into a CFG-local representation. So IMO it's feasible to assume all these variants as well as the local variant to have the exact same number of decimals.

Reply:
Definitely need to keep that for now. Otherwise, we possibly can get mismatches of amounts.